tls/s2n_crl.c

```diff
@@ -0,0 +1,85 @@
+/*
+* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include "s2n_crl.h"
+#include "tls/s2n_connection.h"
+
+static struct s2n_x509_crl* s2n_x509_crl_new(void) {
+    DEFER_CLEANUP(struct s2n_blob mem = { 0 }, s2n_free);
+    PTR_GUARD_POSIX(s2n_alloc(&mem, sizeof(struct s2n_x509_crl)));
+
+    struct s2n_x509_crl *crl = (struct s2n_x509_crl*)(void*) mem.data;
+    crl->crl = NULL;
+
+    ZERO_TO_DISABLE_DEFER_CLEANUP(mem);
+    return crl;
+}
+
+int s2n_x509_crl_from_pem(uint8_t *pem, size_t len, struct s2n_x509_crl **crl) {
+    POSIX_ENSURE_REF(crl);
+
+    struct s2n_blob pem_blob = { 0 };
+    POSIX_GUARD(s2n_blob_init(&pem_blob, pem, len));
+
+    struct s2n_stuffer pem_stuffer = { 0 };
+    POSIX_GUARD(s2n_stuffer_init(&pem_stuffer, &pem_blob));
+    POSIX_GUARD(s2n_stuffer_skip_write(&pem_stuffer, pem_blob.size));
+
+    DEFER_CLEANUP(struct s2n_stuffer der_out_stuffer = {0}, s2n_stuffer_free);
+    POSIX_GUARD(s2n_stuffer_growable_alloc(&der_out_stuffer, 2048));
+    POSIX_GUARD(s2n_stuffer_crl_from_pem(&pem_stuffer, &der_out_stuffer));
+
+    DEFER_CLEANUP(struct s2n_blob crl_blob = { 0 }, s2n_free);
+    POSIX_GUARD(s2n_alloc(&crl_blob, s2n_stuffer_data_available(&der_out_stuffer)));
+    POSIX_GUARD(s2n_stuffer_read(&der_out_stuffer, &crl_blob));
+
+    *crl = s2n_x509_crl_new();
+    POSIX_ENSURE_REF(*crl);
+
+    const uint8_t *data = crl_blob.data;
+    (*crl)->crl = d2i_X509_CRL(NULL, &data, crl_blob.size);
+    POSIX_ENSURE((*crl)->crl != NULL, S2N_ERR_INTERNAL_LIBCRYPTO_ERROR);
```

Comment:
Why allocate more space for the same data / make another copy (crl_blob)? Why can't you use the der_out_stuffer's blob directly, or do a [raw read](https://github.com/aws/s2n-tls/blob/main/stuffer/s2n_stuffer.h#L105)?

Reply:
Right yeah. I changed this to use the der_out_stuffer's blob directly.